### PR TITLE
[datadog] remove agent5 support, cleanup the chart

### PR DESCRIPTION
**Pull request by xvello**
This PR is part one of the Datadog chart revamp, aiming at improving its maintainability. It brings the following changes:

- deploying Agent5 (`datadog/docker-dd-agent` image) is not supported anymore, users can still run it with previous versions of the chart. An error is displayed during install, with pointers to the 0.18.0 chart version. Agent6 has been released 4 months ago
- remove agent5-only elements (legacy trace port, agent5 envvars, autoconf configs mount) from both daemonset and deployment
- sync the clusterrole with the [upstream one](https://github.com/DataDog/datadog-agent/blob/master/Dockerfiles/manifests/rbac/clusterrole.yaml), remove permissions only needed by agent5
- soft-deprecate the `autoconf` option and it's associated configmap:
  - merge it's content in the `confd` configmap for compatibility (in case of name collisions, the `confd` map takes precedence)
  - print a deprecation notice during installation if the option is found
- don't create the `confd` and `checksd` configmaps when empty
- cleanups and rewordings in the `README.md` and example `values.yaml`
- update tag to `6.3.2`
- add a deprecation notice on install if both daemonset+deployment are enabled, pointing to the leader election feature, and asking users to split in two separate releases if they still need this deployment case

### CODE REVIEW

**Comment by xvello** [comment]
@unguiculus as this includes deprecations, I bumped the major version. I'd love to have your input on this process.

File: stable/datadog/templates/NOTES.txt [hunk 1]
```diff
@@ -28,5 +28,47 @@ Then run:
...
+
+The autoconf value is deprecated, Autodiscovery templates can now
+be safely moved to the confd value. As a temporary measure, both
+values were merged into the {{ template "datadog.confd.fullname" . }} configmap,
+but this will be removed in a future chart release.
+Please note that duplicate file names may have conflicted during
+the merge. In that case, the confd entry will take precedence.
+{{- end }}
+
+{{- if eq .Values.image.repository "datadog/docker-dd-agent" }}
```
Comment (devonboyer) [inline stable/datadog/templates/NOTES.txt:65]: If I use `--set image.repository=datadog/docker-dd-agent` I will see both the deprecation notice and the error message. It seems confusing to have both of these?

**Review by devonboyer — COMMENTED** [review]
Tested:

`helm install --set datadog.apiKey=<API_KEY> .`

`helm install --set datadog.apiKey=<API_KEY> --set image.repository=datadog/docker-dd-agent .`

I wasn't sure how to test that --version 0.18.0 still works as intended?

A few questions/comments:

- `appVersion` should be changed to `6.3.1` in `Chart.yaml` to match the default tag in `values.yaml`
- Does it make sense to allow the user to choose whether or not to mount the `dockersocket` or use the kubelet IP?
- Did we decide that the user needs to manually create the `datadogtoken` configmap for events collection?

File: stable/datadog/templates/NOTES.txt (hunk 1 above)
Comment (xvello) [reply stable/datadog/templates/NOTES.txt:65]: Good catch, the condition for the autoconf warning was bad. Fixing it

**Comment by xvello** [comment]
Thanks @devonboyer 

> I wasn't sure how to test that --version 0.18.0 still works as intended?

IIRC the `--version` flag works when using the helm repo, but not a local checkout. I confirmed running `helm install --version 0.17.0 stable/datadog` successfully installs a previous version

>appVersion should be changed to 6.3.1 in Chart.yaml to match the default tag in values.yaml

👍 

>Does it make sense to allow the user to choose whether or not to mount the dockersocket or use the kubelet IP?

We'll need to add an option to support several container runtimes, adding a card to the backlog once we're clearer on what we need to mount for containerd.

>Did we decide that the user needs to manually create the datadogtoken configmap for events collection?

Made a backlog card for that to address later. For now, that configmap name is non configurable, so it would collide across helm installs.

**Comment by irabinovitch** [comment]
As this is a fairly major change to the chart targeted for a major release, it would be good to get some feedback from other contributors on the project.

CC: @cw-sakamoto @timurkhafizov @knqyf263  @MasayaAoyama @snebel29 @flah00  @LiranLast  @kernelpanek  @vreon  @ivanilves  @gtaylor  @ivanfitenko  @c-knowles @awwithro @AdamDang

**Comment by vreon** [comment]
@irabinovitch LGTM

**Comment by ivanilves** [comment]
LGTM from my side! 👍🏻

**Comment by LiranLast** [comment]
Looks good. I feel it would be useful to be able to expose the DaemonSet as a service as well.

While the statsd SDK supports reaching out to the Node's IP, the APM SDK does not.

**Comment by xvello** [comment]
Thanks @knqyf263 @vreon @c-knowles @ivanilves @LiranLast for your feedback!

@c-knowles [Unix sockets for dsd](https://docs.datadoghq.com/developers/dogstatsd/unix_socket/) was introduced to handle these reachability issues, did you roll back to UDP, or do you still have some un-migrated apps?

The alternative would be to run a second release of the chart, as a daemonset of the `datadog/dogstatsd` image, with the `DD_ENABLE_METADATA_COLLECTION` envvar set to `false`, and host port enabled (or a daemonset with a service, if hostPort and services do not play nice in your CNI). Am I missing another requirement?

@LiranLast The goal of removing simultaneous daemonset+deployment is to refactor the templates into one, and merge both feature sets. The next chart version will support optionally exposing the daemonset's port as a service, although host port will still be the preferred way, in order to keep traffic local.

**Comment by c-knowles** [comment]
@xvello we need UDP for sure, not everything supports unix socket and some apps likely never will as they support stats not dogstatsd. We still want to link them with the downward API to the local node but it'll take some time to update them all. So in the meantime we need some minimal install for k8s Service backed by dogstatsd. It doesn't make much sense for us in this combination to have that as a daemon as we only need a few pods per cluster. I'm not sure of how well maintained the dogstatsd image is as agent v6 was not so well documented for that mode when we checked a few weeks ago plus https://github.com/DataDog/datadog-agent/issues/1640 isn't fixed yet.

**Comment by vreon** [comment]
@LiranLast @xvello I forgot to mention we also create a Service for the DaemonSet at my org, so I agree that'd be nice to have in the chart. I think `externalTrafficPolicy: Local` on that would keep traffic node-local, but I haven't had a chance to try it.

**Comment by c-knowles** [comment]
@xvello great, thanks you! FYI our internal ticket ref is 144293.

**Comment by xvello** [comment]
Thanks everyone for your feedback, let's focus the efforts for exposing the service on https://github.com/kubernetes/charts/pull/6615

I will rebase and test this branch before merging it

**Review by hkaj — APPROVED** [review]
Nice one, thanks a lot for this one @xvello, it's cool to see it happening!

**Comment by mausch** [comment]
Not sure if this is the right place to post it, but ever since upgrading my chart to 1.0.0 the datadog pods have been restarting like crazy.

Using datadog/agent:6.4.1 I get:

```
[ AGENT ] 2018-08-07 13:07:24 UTC | WARN | (start.go:273 in StopAgent) | Some components were unhealthy: [collector-queue]
[PROCESS] process-agent exited with code 0, disabling
[ AGENT ] AGENT EXITED WITH CODE 0, SIGNAL 0, KILLING CONTAINER
```

with 6.3.3 and 6.2.1 I get similar behaviour but no hints as to why it exits.

I just rolled back to chart 0.10.10 with datadog/agent:6.4.1 and it works fine again.

**Comment by xvello** [comment]
Hi @mausch 

An unhealthy `collector-queue` happens when the collector cannot run all scheduled checks in time. While we are working for a more reliable collector in 6.5.0, our recommendation to handle this is to raise the value of `DD_CHECK_RUNNERS` (2 by default). If the issue persists, please contact support to investigate if a check has a performance regression.

At first sight, the only chart change that could influence this is merging the legacy `autoconf` array, while previous chart versions ignored it. That might lead to additional checks running with `1.0.0`

**Pull request merged**